### DISCUSSION TITLE
JS SSR - issues and recommendations when interacting with PocketBase

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#5313] [Announcements]
This is more of rant as a result of the many JS SSR opened discussions and a recent security report that I don't think I managed to explain to the user properly, so I'll attempt to do it here with little more details.

## The elephant in the room

If you are using SvelteKit, Nuxt, Next.js, etc. in a JS SSR mode, **you already have a backend and you might not need PocketBase or other external service**. There are plenty of JS libraries that you could combine together to build your application. For example, if you need something to take care for the authentication you could check [Lucia](https://github.com/lucia-auth/lucia) or [NextAuth.js](https://github.com/nextauthjs/next-auth); for interacting with the database you can make use of [Drizzle](https://orm.drizzle.team/)/[Prisma](https://www.prisma.io/) (_they both support SQLite and comes with a helper UI for simple data explorations_), and so on.

## The problems of PocketBase with JavaScript meta frameworks

JavaScript meta frameworks like SvelteKit, Nuxt, Next.js, etc. are very versatile and could run in various modes.
Some are easier to integrate with, others - not much due to their specific rendering design.

When interacting with an external services like PocketBase from a meta JavaScript framework you usually have 3 options:

1. only from client-side (ex. browser->PocketBase)
2. only from server-side (ex. browser->node->PocketBase)
3. mixed - from both client-side and server-side

With risk to make the matter slightly more complicated, it also depends how the JavaScript server-side code is being executed - as a long running server process(es) or in a "serverless"/worker/lambda mode (ala PHP as short-lived request-response cycles).

Based on the PocketBase reported issues so far, most problems and complications arise with 2) and 3) and when the JS server code is being executed as a long running process.

### Problem 1 - Global/shared PocketBase JS SDK server instance

While the underlying issue is not PocketBase specific, this is the most common issue and the most serious one since it introduces a security vulnerability in your app (and it is very easy to slip into production because rarely developers test their applications with concurrent requests from different users locally).

If you have the following **server-side** JS pseudo-code:
  ```js
  import PocketBase from "pocketbase"

  const pb = new PocketBase(...)

  export default pb // unsafe for regular user auth, instead should export a factory function

  // ... in some other file ...

  import pb from "./pb.js"

  async function serverAction(req, resp) {
      await pb.collection("users").authWithPassword(...)
  }
  ```

Then this is vulnerable because requests from different users will overwrite and modify the same `pb.authStore` object.
To solve it, **you have to instantiate a new SDK instance and keep it scoped within the duration of the individual server request**. Most JS frameworks usually have `req.locals` or other similar object that could be used to safely store request related data. Unfortunately there are exceptions to this, for example Next.js, that don't offer easy sharing complex request data between handlers and middlewares, and for such cases you usually have to create a new separate SDK instance inside each different handler/middleware.

**It is safe to have a single PocketBase JS SDK instance only when used:**
  - on the client-side (ex. browser)
  - on the server-side when in "serverless"/worker/lambda/etc. mode that essentially init a new js process with each request and destroy it after serving the response
  - on the server-side when used as admin and you are not keeping track of regular user auth state on the server

### Problem 2 - OAuth2 integration

Most OAuth2 reported issues come from misunderstanding how the current OAuth2 "all-in-one" flow works.
When calling `authWithOAuth2({...})` we start a one-off realtime subscription and open a new window/tab with the provider's login page.

Once the login has been completed, the provider notify the PocketBase server and we send back to the client the received data over the previously started realtime connection (aka. the page that has started the connection **must be still active**).

This flow works fine when used from the client-side (e.g. in a browser or mobile app) but fails when it is executed in a _server-only_ action because, as mentioned above, the server needs to be able to open a browser window and to keep the connection with the PocketBase server active.

If you want a _server-only_ OAuth2 integration you need to use the "Manual code exchange" flow (the second tab in the [OAuth2 integration guide](https://pocketbase.io/docs/authentication/#oauth2-integration)).

An alternative if you don't want a _node-only_ access to your PocketBase server is to create a dedicated client-side JS SDK instance, call `authWithOAuth2()` from the client and then add the result as cookie via `document.cookie = pb.authStore.exportToCookie()` so that your Node server could receive the auth state on the next server request.

### Problem 3 - Mixed client and server access

When you want to keep the user auth state and interact with PocketBase both from the browser and Node you have to maintain 2 type of SDK instances - one unique for each server request and another (usually global) for the client.

To sync and share the auth state between both environments I usually recommend using cookies because browsers transparently will send it to the server with each request. The caveat here is that in order to have access to the cookie from the client instance you'll have to export it with `httpOnly:false`. This is safe as long as you use only trusted 3rd party JavaScript (if at all) and have a basic [Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) in place either as meta tag or HTTP header.

### Problem 4 - API cache issues and custom `fetch` functions

PocketBase doesn't have API response cache out of the box but some meta frameworks like Next.js have non-standard `fetch` implementation that comes with cache enabled by default. While this is usually extensively explained in [their docs](https://nextjs.org/docs/app/building-your-application/caching), most users don't seem to read it or understand the caveats that comes with it and it ended up being reported as PocketBase issue (**sometimes it could be even a security one**). Fortunately this seems that will change with Next.js 15 but still meta frameworks often use their own `fetch` and it will affect the JS SDK.

**By default the JS SDK uses whatever global `fetch` implementation is available so all options supported by the custom `fetch` function are also supported by the JS SDK**.

For the cases where you need to set an entire custom `fetch` function per request basis (ex. as in [SvelteKit `load`](https://kit.svelte.dev/docs/load#making-fetch-requests)), you can pass the custom fetch function as a prop to the optional `SendOptions` object that all requests support (usually the last optional argument). For example:

  ```js
  await pb.collection("example").getList(1, 30, {
      expand: "something",
      fetch: yourCustomFetchFunc,
  })
  ```

### Problem 5 - Proxying realtime connections

If you want PocketBase to be accessible **only** from your JS server app and want to use the realtime subscriptions, then you'll have to maintain on your own a separate persistent connection between the client and your node server and proxy the related data back to the client. This most of the time is understood by the users but what usually is reported as an issue is the delay when users receive the event for a changed record for example.
This delay could be caused by many reasons (including poor/inefficient implementation of the user's JS server SSE handling) but often the culprit ended being the data buffering that some frameworks have by default. For regular proxies like Nginx we return a special `X-Accel-Buffering:no` header that instruct the proxy to disable it but this is not a standard header and devs sometimes need to spend a little more time configuring their JS server and disable buffering on their own.

### Problem 6 - Performance and deployment

Another common issue that most users seems to ignore until they hit production is the introduced performance bottleneck caused by proxying every request through their single threaded Node.js server. Yes, cluster/workers mode, pm2, etc. exists but it is one more thing that you'll have to configure, deploy and maintain.
Additionally, often if you want to host your JS SSR application and PocketBase on the same server, you will need a reverse proxy like Nginx, Caddy, Apache which could further complicate the deployment. The less services you have to maintain on the server the better.

## Final notes

The above problems are the most common ones that users ended up reporting but there are probably others (ex. accidentally using a server variable inside a client-side component).

A lot of modern meta-frameworks like Next.js, SvelteKit, Nuxt seems to encourage JS SSR first approach, but **I think this is a poor choice especially when you have to integrate with external backend like PocketBase and you are not working on a website**.

For the majority of the cases a **client-side rendered app will be many times more simple and performant than a JS SSR one**.

Client-side rendered apps provide a clear distinction between frontend and backend, there are easier to test and reason when/where your code is being executed.

What is so wrong for example with SPAs that most users seems so against it (concerns for accessibility and bundle size/split are pretty much resolved nowadays with modern bundlers like Vite)?
Are you really targeting users that disable JavaScript?
Have you ever tested your JS SSR application whether it works with disabled JavaScript at all?

_"But Gani I want to use PocketBase with JS SSR"_.

Fine, use it. There is nothing wrong as long as you understand the trade-offs that comes with that choice.
No one is going to stop you (or care what you are using), but don't expect everything to work out of the box and don't wait on me to fix your security issue just because you copied it from some random tutorial.

Supabase for example has a dedicated [`@supabase/ssr`](https://www.npmjs.com/package/@supabase/ssr) package to assist with the JS SSR integrations but I neither have the time, nor I want to maintain something that personally I'll probably never use (nor I could justify the cost of hiding the JS SSR complexity with another layer of abstraction).

As mentioned in several other discussions, the easiest and most efficient way at the moment to use PocketBase is by interacting with it directly from the client-side. That is why we have the builtin JSON APIs and API rules in the first place.

It is fine to have your own different preferences and opinions. Don't be "absolutist" and understand that every external dependency, library, framework that you bring is a trade-off.

My _personal_ preference is a SPA (usually plain Svelte + hash based router, for example [`presentator/presentator`](https://github.com/presentator/presentator)) and this is what I usually recommend because I find it easier to work with. If you need additional server-side handling (to handle payment webhook, extra data server validations, etc.) you could:
  - Intercept existing or create new APIs by [using Pocketbase as framework and extend with Go/JS](https://pocketbase.io/docs/use-as-framework/) (_with the next release the Go APIs should be even more simple and intuitive to use_).

  - If you still want to use JS on the server-side, then you could create one-off node server-side actions that will interact with PocketBase only as admin/superuser and as pure data store (this is like regular database access but over HTTP).
    To initialize an admin client it is enough to authenticate as admin (in this case it is also safe to have a global JS SDK instance). For example:
    ```js
    // src/superuser.js
    import PocketBase from "pocketbase"

    const superuserClient = new PocketBase('https://example.com');

    // disable autocancellation so that we can handle async requests from multiple users
    superuserClient.autoCancellation(false);

    // option 1: authenticate as superuser using email/password (could be filled with ENV params)
    await superuserClient.collection('_superusers').authWithPassword(SUPERUSER_EMAIL, SUPERUSER_PASS, {
      // This will trigger auto refresh or auto reauthentication in case
      // the token has expired or is going to expire in the next 30 minutes.
      autoRefreshThreshold: 30 * 60
    })

    // option 2: OR authenticate as superuser via long-lived "API key"
    // (see https://pocketbase.io/docs/authentication/#api-keys)
    superuserClient.authStore.save('YOUR_GENERATED_SUPERUSER_TOKEN')

    export default superuserClient;
    ```

    Then you can just import the file in your server-side components and use it as usual:
    ```js
    import superuserClient from './src/superuser.js'

    async function serverAction(req, resp) {
      ... do some extra data validations or handling ...

      // send a create request as superuser
      await superuserClient.collection('example').create({ ... })
    }
    ```

With that said, after the upcoming v0.23.0 release (_there are still no hard ETAs but expect release within the next 1-2 months at worst_), I'll add various helpers for **Go SSR** apps because I personally have a need for that and I'm planning creating a new simple Go template engine (and eventually "pb_pages" files based router as a PocketBase plugin), but this remains a low priority for now and users will have to create these helpers on their own if they want to use PocketBase for building SSR apps.

I hope the above clear some misunderstanding and please refrain from heated arguments. Again, I have nothing against JS SSR as long as you understand the drawbacks when used with PocketBase. Do whatever you think best suit your project and move on.

_edit: updated the superuser example_

**Comment by RenFontes** [comment]
Thanks for writing this, I'll have to check my app to see that I didn't introduce any security issues related to SSR. 🤔

Last year I created a small site using pocketbase and sveltekit, now I'm wondering how I did it because most of it just worked I think.

BTW, thanks for all your hard work! You are amazing and I admire you a lot. I wish more people would know the wonders of Pocketbase, how good it is and how cheap it is to deploy compared to firebase or supabase. :))

**Comment by vigneshksaithal** [comment]
I am using SvelteKit, should I be worried? Or is it for a certain edge case?

**Reply by ganigeorgiev** [reply]
If you have followed the JS SDK ["SSR integration"](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration) README example exactly as shown then you are fine, but again as mentioned in "Problem 1" above, you need a **new JS SDK instance per each new request** otherwise requests from different users could end up overwriting and modifying the same object.

This is not specific only to the PocketBase JS SDK and apply for any shared/global objects.

In the example provided in "Problem 1" the exported `pb` instance is such shared object because **the imported file is evaluated only once** no matter in how many different files you import it. That's why the recommendation in this case is to return a "factory" function, aka. function that will create new PocketBase instance inside each server action.

I really don't know how to explain it better and I feel that the more I try to clarify the caveats the more complicated I make it look.

**Reply by Khalilsqu** [reply]
i believe in sveltekit, the best is to have the pocketbase instance inside sveltekit locals. in a server hook


```
// src/hooks.server.js

import PocketBase from 'pocketbase';

export const handle = async ({ event, resolve }) => {
	event.locals.pb = new PocketBase('http://localhost:8090');
	// whatever other logic here !!
	const response = await resolve(event);
	return response;
};
```

then you can access the locals in sveltekit loaders or actions like below

```
export const actions = {
	login: async (event) => {
        const userAuthTokenValid = event.locals.pb.authStore.isValid
```

**Reply by vigneshksaithal** [reply]
I access PocketBase via the hooks locals as suggested in the docs. So no problem I think.

**Reply by ZeroPie** [reply]
I have 

```
import PocketBase from 'pocketbase';
import { PUBLIC_POCKETBASE_URL } from '$env/static/public';

export function createInstancePbUserInstance() {
	return new PocketBase(PUBLIC_POCKETBASE_URL);
}

```

```
import { login } from '$lib/utils/routes';
import { lucia } from '$lib/server/auth';
import { redirect } from '@sveltejs/kit';
import PocketBase from 'pocketbase';
import { PUBLIC_POCKETBASE_URL } from '$env/static/public';
import { createInstancePbUserInstance } from '$lib/server/db';
import { SECRET_POCKETBASE_ADMIN, SECRET_POCKETBASE_ADMIN_PASSWORD } from '$env/static/private';

const isProtectedRoute = (routeId: string | null) => routeId?.startsWith('/(private)');

export const pbAdmin = new PocketBase(PUBLIC_POCKETBASE_URL);
await pbAdmin.admins.authWithPassword(SECRET_POCKETBASE_ADMIN, SECRET_POCKETBASE_ADMIN_PASSWORD);

export const handle = async ({ event, resolve }) => {
	const sessionId = event.cookies.get(lucia.sessionCookieName);
	// edited:
	event.locals.pbUser = createInstancePbUserInstance();
	event.locals.pbAdmin = pbAdmin;
```

it helps to be mindful of wether you use a pb instance that's authenticated as an admin
or a pbinstance that's authenticated as a user (although in my case im about to remove it since im using lucia) but that one does indeed create a new instance on each request

**Reply by ganigeorgiev** [reply]
@ZeroPie I'm not sure that I understand the above code sample, but note that the import:

```js
import { pbUser } from '$lib/server/db';
```

**doesn't create a new instance for each request**. It reuses the one created in the imported file for the same reason explained in the [comment above](https://github.com/pocketbase/pocketbase/discussions/5313#discussioncomment-10253779), aka. the line `export const pbUser = createInstance()` would be evaluated only once and the line `event.locals.pbUser = pbUser` is vulnerable. It should be `event.locals.pbUser = createInstance()` (or just `event.locals.pbUser = new PocketBase(PUBLIC_POCKETBASE_URL)`).

As mentioned in the post, it is safe to have a global/singleton JS SDK instance only:

- on the client-side (ex. browser)
- on the server-side when in "serverless"/worker/lambda/etc. mode that essentially init a new js process with each request and destroy it after serving the response
- on the server-side when used as admin and you are not keeping track of regular user auth state on the server

**Reply by ZeroPie** [reply]
Noted.

This is indeed what i meant to do
```event.locals.pbUser = createInstance()```
but my attention failed me.

Thank you!
I edited my previous snippet as to make it easier for future readers to see the correct version right away.

**Comment by seantiz** [comment]
Yeah I am guilty of using the db singleton in testing, so thanks for running through this issue. In fairness, the last time I did it, it was a read-heavy app with no writes to the db from end users. But I still have to shake the habit.

**Comment by blackfyre** [comment]
This is the reason for why I went with htmx. Thanks for summing it up!

**Reply by Sleepful** [reply]
Going this htmx go/ssr route is not without its difficulties.
I did it on a sample project, others may use it as reference, it shows how to do it and also the messy code that it creates with the current PB version (0.26):

https://github.com/Sleepful/mymail/tree/main/app/router

also see: https://github.com/pocketbase/pocketbase/discussions/5307

**Comment by MrIceman** [comment]
Now I am not using Pocketbase but it seems like the ones who complain about the mentioned security issues failed to understand simple principles such as your backend should be stateless. Funny how a JS developer reports security issues because he doesn't understand that state should be scoped to a request

**Comment by jeffscottward** [comment]
Ok so client-side Next.JS only then.

1) I like that front/back distinction as you mentioned.
2) It's how I've been doing Next for 7 years, I never really liked this JS SSR push.
3) I'm building an agentic project so a frontend isn't even necessary meaning I get way more value out of Pocketbase than I do Next.js for this.
4) Security isn't my strong suit so I don't even wanna go near all that.

Excellent write up and thank you for everything you do.

I do have a request (maybe I should file this), it's an AI LLM to work with PB.

I found https://postgres.new absolutely fantastic, but now have to convert to SQLite to bring the schema in.

Wondering if being able to do operations with AI right in the interface leaning on Chrome's window.AI would be reasonable?

**Comment by lumi-mustafa-hanif** [comment]
Amazing write-up, thank you for the work you do.  

You hit the nail on the head with this recommendation:

> If you still want to use JS on the server-side, then you could create one-off node server-side actions that will interact with PocketBase only as admin/superuser and as pure data store

I use it just exactly like this, my main value from PB is its admin panel and file upload, I didn't know Prisma Studio was a thing, I'll explore if it has file upload. May need to move towards that.

Also I noticed that `expand` queries sometime randomly don't return anything in the expand object, but I guess thats a separate issue.

**Reply by ganigeorgiev** [reply]
> Also I noticed that expand queries sometime randomly don't return anything in the expand object, but I guess thats a separate issue.

There is nothing random in the `expand` behavior. It should resolve the relations as long as your client request satisfies the View API rule of the expanded collection(s). If you use Next.js make sure that what you are experiencing is not caused by its default cache.

**Comment by otowa1** [comment]
This is a very good explanation. 
By the way, is this the right link?

> My personal preference is a SPA (usually plain Svelte + hash based router, for example [presentator/presentator](https://github.com/presentator/presentator))

The linked application does not look like a hash based router.

**Reply by ganigeorgiev** [reply]
@otowa552 The linked repo was intended to serve as an example for "SPA + hash based router" PocketBase application.

Under the hood it utilizes [ItalyPaleAle/svelte-spa-router](https://github.com/ItalyPaleAle/svelte-spa-router) _(it has some quirks but I like this specific library because is tiny and the code is easy to read and follow)_.

**Reply by otowa1** [reply]
Thanks!
I'm embarrassed by the mistake I wouldn't make if I could interpret the text properly.
As an excuse, I was able to interpret the text properly when I read it before, but recently I got a new job and was too tired to think straight.
I was wondering about using the famous routing libraries in this case, as many of them are heavyweight, so it is very helpful to know the recommended configuration.

**Comment by FoolsTech** [comment]
The only reason I need Nuxt/SSR for is SEO. For a SaaS I wouldn't need it. But I thought about using pocketbase for a blog as well... hmm...

**Reply by ganigeorgiev** [reply]
I'm not saying that _"You should never use JS SSR"_, just that it comes with a lot of caveats when integrating with PocketBase and that developers need to carefully evaluate whether the potential gains would outweigh the cost of the integration for their specific use case.

**Reply by FoolsTech** [reply]
Yes, I got it, you describe general JS SSR considerations, regardless the backend tool/framework. I just never thought about these issues since I never run Nuxt in production, yet. 

Would you be open to for pull requests in this regard or should the stuff rather go in a separate repo/project should I come up with some solutions? Cause I have been working with Pocketbase last month full time and want to use it for everything now :D It made a huge shift for me in approaching indie hacking / MVPs. I don't want to touch anything else for web projects :D

**Reply by ganigeorgiev** [reply]
After v0.23.0 release there are plans to add "Guides/Examples" section on the website where community contributions would be helpful but for now it is better such external integrations and examples to live in their own repo/gist (of course feel free to link it here or in a ["Show and tell" discussion](https://github.com/pocketbase/pocketbase/discussions/categories/show-and-tell)).

**Reply by FoolsTech** [reply]
cool, thanx!

**Comment by kuwts** [comment]
If I am using Astro with SSR, should I be worried? 

I would love to use only a client side JS framework but I went with Astro for the easy page routing and SEO advantages. I could handle auth, CRUD, etc requests using Svelte components inside Astro but then would still need to check auth status to access user dashboards, verification, etc with Astro anyway.

**Reply by ganigeorgiev** [reply]
As long as you take in account the above mentioned problems and follow the [Astro SSR minimal example](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration) from the JS SDK readme for keeping track of the regular user auth state, then it should be fine.

**Reply by jeffscottward** [reply]
> If I am using Astro with SSR, should I be worried? 
> 
> 
> 
> I would love to use only a client side JS framework but I went with Astro for the easy page routing and SEO advantages. I could handle auth, CRUD, etc requests using Svelte components inside Astro but then would still need to check auth status to access user dashboards, verification, etc with Astro anyway. 

I'm going to be working with the same stack, please keep us updated. Or me on twitter @jeffscottward

**Comment by ikoral** [comment]
GitHub Oath2 login flow needs the browser so you need to call from a client component like 
```
const authData = await pbClientInstance.collection('users').authWithOAuth2({ provider: 'github' });
```
then you could make an API call to your Next.js route to set an HTTP-only cookie. You don't need to manage any other API in Next.js this is the only exception you need to set up in your Next.js.

your client component:
```
   const pb = createBrowserClient();
    try {
      const authData = await pb
        .collection("users")
        .authWithOAuth2({ provider: "github" });

      const { token, record } = authData;

      // Store the token in an HttpOnly cookie (send your Next.js route)
      await fetch("/api/set-cookie", {
        method: "POST",
        body: JSON.stringify({ token, record }),
      });
```

your Next.js route
./src/app/api/set-cookie/route.tsx
```
import { cookies } from "next/headers";
import { NextResponse } from "next/server";

export async function POST(request: Request) {
  const authData = await request.json();
  const { token, record } = authData;
  const cookie = JSON.stringify({ token, record });

  (await cookies()).set("pb-auth", cookie, {
    httpOnly: true,
    secure: process.env.NODE_ENV === "production",
    sameSite: "lax",
    maxAge: 60 * 60 * 24 * 7, // 1 week
  });

  return NextResponse.json({ success: true });
}
```

**Reply by mihaiandrei97** [reply]
@ikoral Is it safe to have an endpoint that allows anyone calling it to set a http only cookie?

**Reply by ikoral** [reply]
Hi @mihaiandrei97, if I miss some points please let me know This is my 5 cents:
Anyone can set any cookie in their browser even using browser dev tools. Cookie content matters here. What content do you set an HTTP-only cookie with? Where do you use that cookie if you made up its content? You need to log in first and get the correct cookie then you send your endpoint (of course it should be secure TLS/SSL aka https:// endpoint.) with payload. You might think about limiting requests for that endpoint. 

Again if I am missing something please let me know.

**Comment by bartonhammond** [comment]
I followed the `superuser.js` approach with my Astro project which has no users - no authentication.  I'm just doing simple CRUD with one collection `contacts`.  I'm using PB from my NodeJS server code.   

I end up getting an exception here but it's empty

```
  import pb from "./superuser.js"
.....
  try {
    let results = await pb.collection.getList(page, limit, options);
    console.log(`getContacts: results: ${results}`)
    contacts = results.items;
          console.log(`getContacts: contacts: ${contacts}`)  
  } catch (e) {
    console.log(`getContacts error: ${JSON.stringify(e)}`);  //<<<<< I get this exception but its empty
  }
  return contacts;
```
Here is the log of the empty exception
```
getContacts error: {}
```

The PB log looks like it's fine?


![image](https://github.com/user-attachments/assets/f81847a2-88ae-4980-aa23-92f1d3824bd0)


If I remove the `API rule` requiring `superuser` it works fine.

**Reply by ganigeorgiev** [reply]
@bartonhammond Without clear steps and more complete code sample to reproduce the error I cannot help you.

Note that calling `JSON.stringify(error)` is not guaranteed to print all error fields. For more details about the available error fields, please check https://github.com/pocketbase/js-sdk?tab=readme-ov-file#error-handling.

To rule out PocketBase issue, you can create a plain `pb_public/index.html`:

```html
<!DOCTYPE html />
<html>
<head>
    <meta charset="utf-8" />
    <title>Test</title>
</head>
<body>
    <script type="module">
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.es.mjs"

        const pb = new PocketBase("http://127.0.0.1:8090");

        ...
    </script>
</body>
</html>
```
(_open the browser and navigate to `http://127.0.0.1:8090`_)

If that doesn't help, please open a new Q&A discussion, but try to provide more context about the problem so others can replicate it.

**Reply by ganigeorgiev** [reply]
@bartonhammond Note also that the call `pb.collection.getList(..)` is invalid. It must be `pb.collection("yourCollectionName").getList(...)`.

But in all cases, please refer to https://github.com/pocketbase/js-sdk?tab=readme-ov-file#error-handling to see the available error fields.

**Comment by bartonhammond** [comment]
Doh!  Sorry about that - I made that change earlier and forgot about it -
yes - putting the "contacts" back in makes all good now.

**Comment by VictorioBerra** [comment]
> I'll add various helpers for Go SSR apps because I personally have a need for that and I'm planning creating a new simple Go template engine (and eventually "pb_pages" files based router as a PocketBase plugin)

Is there any updates for any of this?

**Reply by ganigeorgiev** [reply]
No.
